Plaintext version of logging sample
```
 2024-07-19 02:11:59.374+08:00 WARN (leaderCheckpointer|115) [AuthenticationProviderFactory.installPlugin():32] Plugin MYSQL_NATIVE_PASSWORD has already been installed!
 2024-07-19 02:11:59.375+08:00 WARN (leaderCheckpointer|115) [AuthenticationProviderFactory.installPlugin():32] Plugin AUTHENTICATION_LDAP_SIMPLE has already been installed!
 2024-07-19 02:11:59.377+08:00 WARN (leaderCheckpointer|115) [AuthenticationProviderFactory.installPlugin():32] Plugin AUTHENTICATION_KERBEROS has already been installed!
 2024-07-19 02:11:59.383+08:00 WARN (Thread-42|71) [TemporaryTableCleaner.runAfterCatalogReady():87] listSessions return error from 127.0.0.1:29020, skip clean temporary tables
 org.apache.thrift.transport.TTransportException: java.net.ConnectException: Connection refused (Connection refused)
        at org.apache.thrift.transport.TSocket.open(TSocket.java:231)
        at com.starrocks.common.GenericPool$ThriftClientFactory.create(GenericPool.java:148)
        at com.starrocks.common.GenericPool$ThriftClientFactory.create(GenericPool.java:133)
        at org.apache.commons.pool2.BaseKeyedPooledObjectFactory.makeObject(BaseKeyedPooledObjectFactory.java:62)
        at org.apache.commons.pool2.impl.GenericKeyedObjectPool.create(GenericKeyedObjectPool.java:1036)
        at org.apache.commons.pool2.impl.GenericKeyedObjectPool.borrowObject(GenericKeyedObjectPool.java:356)
        at org.apache.commons.pool2.impl.GenericKeyedObjectPool.borrowObject(GenericKeyedObjectPool.java:278)
        at com.starrocks.common.GenericPool.borrowObject(GenericPool.java:105)
        at com.starrocks.rpc.FrontendServiceProxy.call(FrontendServiceProxy.java:33)
        at com.starrocks.server.TemporaryTableCleaner.runAfterCatalogReady(TemporaryTableCleaner.java:70)
        at com.starrocks.common.util.FrontendDaemon.runOneCycle(FrontendDaemon.java:72)
        at com.starrocks.common.util.Daemon.run(Daemon.java:107)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
        at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:412)
        at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:255)
        at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:237)
        at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.base/java.net.Socket.connect(Socket.java:609)
        at org.apache.thrift.transport.TSocket.open(TSocket.java:226)
        ... 11 more
```

Json format logging
```
{"instant":{"epochSecond":1721326364,"nanoOfSecond":758709000},"thread":"leaderCheckpointer","level":"WARN","loggerName":"com.starrocks.authentication.AuthenticationProviderFactory","message":"Plugin MYSQL_NATIVE_PASSWORD has already been installed!","endOfBatch":false,"loggerFqcn":"org.apache.logging.log4j.spi.AbstractLogger","contextMap":{},"threadId":119,"threadPriority":5}
{"instant":{"epochSecond":1721326364,"nanoOfSecond":759100000},"thread":"leaderCheckpointer","level":"WARN","loggerName":"com.starrocks.authentication.AuthenticationProviderFactory","message":"Plugin AUTHENTICATION_LDAP_SIMPLE has already been installed!","endOfBatch":false,"loggerFqcn":"org.apache.logging.log4j.spi.AbstractLogger","contextMap":{},"threadId":119,"threadPriority":5}
{"instant":{"epochSecond":1721326364,"nanoOfSecond":759261000},"thread":"leaderCheckpointer","level":"WARN","loggerName":"com.starrocks.authentication.AuthenticationProviderFactory","message":"Plugin AUTHENTICATION_KERBEROS has already been installed!","endOfBatch":false,"loggerFqcn":"org.apache.logging.log4j.spi.AbstractLogger","contextMap":{},"threadId":119,"threadPriority":5}
{"instant":{"epochSecond":1721326364,"nanoOfSecond":775814000},"thread":"leaderCheckpointer","level":"WARN","loggerName":"com.starrocks.catalog.ColocateTableIndex","message":"remove 0 invalid tableid: []","endOfBatch":false,"loggerFqcn":"org.apache.logging.log4j.spi.AbstractLogger","contextMap":{},"threadId":119,"threadPriority":5}
{"instant":{"epochSecond":1721326364,"nanoOfSecond":750087000},"thread":"Thread-42","level":"WARN","loggerName":"com.starrocks.server.TemporaryTableCleaner","message":"listSessions return error from 127.0.0.1:29020, skip clean temporary tables","thrown":{"commonElementCount":0,"localizedMessage":"java.net.ConnectException: Connection refused (Connection refused)","message":"java.net.ConnectException: Connection refused (Connection refused)","name":"org.apache.thrift.transport.TTransportException","cause":{"commonElementCount":11,"localizedMessage":"Connection refused (Connection refused)","message":"Connection refused (Connection refused)","name":"java.net.ConnectException","extendedStackTrace":"java.net.ConnectException: Connection refused (Connection refused)\n\tat java.net.PlainSocketImpl.socketConnect(Native Method) ~[?:?]\n\tat java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:412) ~[?:?]\n\tat java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:255) ~[?:?]\n\tat java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:237) ~[?:?]\n\tat java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[?:?]\n\tat java.net.Socket.connect(Socket.java:609) ~[?:?]\n\tat org.apache.thrift.transport.TSocket.open(TSocket.java:226) ~[libthrift-0.20.0.jar:0.20.0]\n"},"extendedStackTrace":"org.apache.thrift.transport.TTransportException: java.net.ConnectException: Connection refused (Connection refused)\n\tat org.apache.thrift.transport.TSocket.open(TSocket.java:231) ~[libthrift-0.20.0.jar:0.20.0]\n\tat com.starrocks.common.GenericPool$ThriftClientFactory.create(GenericPool.java:148) ~[starrocks-fe.jar:?]\n\tat com.starrocks.common.GenericPool$ThriftClientFactory.create(GenericPool.java:133) ~[starrocks-fe.jar:?]\n\tat org.apache.commons.pool2.BaseKeyedPooledObjectFactory.makeObject(BaseKeyedPooledObjectFactory.java:62) ~[commons-pool2-2.3.jar:2.3]\n\tat org.apache.commons.pool2.impl.GenericKeyedObjectPool.create(GenericKeyedObjectPool.java:1036) ~[commons-pool2-2.3.jar:2.3]\n\tat org.apache.commons.pool2.impl.GenericKeyedObjectPool.borrowObject(GenericKeyedObjectPool.java:356) ~[commons-pool2-2.3.jar:2.3]\n\tat org.apache.commons.pool2.impl.GenericKeyedObjectPool.borrowObject(GenericKeyedObjectPool.java:278) ~[commons-pool2-2.3.jar:2.3]\n\tat com.starrocks.common.GenericPool.borrowObject(GenericPool.java:105) ~[starrocks-fe.jar:?]\n\tat com.starrocks.rpc.FrontendServiceProxy.call(FrontendServiceProxy.java:33) ~[starrocks-fe.jar:?]\n\tat com.starrocks.server.TemporaryTableCleaner.runAfterCatalogReady(TemporaryTableCleaner.java:70) ~[starrocks-fe.jar:?]\n\tat com.starrocks.common.util.FrontendDaemon.runOneCycle(FrontendDaemon.java:72) ~[starrocks-fe.jar:?]\n\tat com.starrocks.common.util.Daemon.run(Daemon.java:107) ~[starrocks-fe.jar:?]\nCaused by: java.net.ConnectException: Connection refused (Connection refused)\n\tat java.net.PlainSocketImpl.socketConnect(Native Method) ~[?:?]\n\tat java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:412) ~[?:?]\n\tat java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:255) ~[?:?]\n\tat java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:237) ~[?:?]\n\tat java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[?:?]\n\tat java.net.Socket.connect(Socket.java:609) ~[?:?]\n\tat org.apache.thrift.transport.TSocket.open(TSocket.java:226) ~[libthrift-0.20.0.jar:0.20.0]\n\t... 11 more\n"},"endOfBatch":false,"loggerFqcn":"org.apache.logging.log4j.spi.AbstractLogger","contextMap":{},"threadId":71,"threadPriority":5}
```

--- a/fe/fe-core/pom.xml
+++ b/fe/fe-core/pom.xml
@@ -392,6 +392,11 @@ under the License.
             <artifactId>log4j-slf4j-impl</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-layout-template-json</artifactId>
+        </dependency>
+
         <!-- https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j-1.2-api -->
         <dependency>
             <groupId>org.apache.logging.log4j</groupId>

--- a/fe/fe-core/src/main/java/com/starrocks/common/Config.java
+++ b/fe/fe-core/src/main/java/com/starrocks/common/Config.java
@@ -98,6 +98,8 @@ public class Config extends ConfigBase {
     @ConfField
     public static boolean sys_log_to_console = false;
 
+    @ConfField(comment = "Log4j layout format. Valid choices: plaintext, json")
+    public static String sys_log_format = "plaintext";
     /**
      * audit_log_dir:
      * This specifies FE audit log dir.

--- a/fe/fe-core/src/main/java/com/starrocks/common/Log4jConfig.java
+++ b/fe/fe-core/src/main/java/com/starrocks/common/Log4jConfig.java
@@ -258,7 +258,53 @@ public class Log4jConfig extends XmlConfiguration {
         if (jsonLoggingConfValue.equalsIgnoreCase(Config.sys_log_format)) {
             // json logging
             String jsonLayout =
-                    "<JSONLayout compact=\"true\" eventEol=\"true\" properties=\"true\" stacktraceAsString=\"true\" />";
+                    "<JsonTemplateLayout locationInfoEnabled=\"true\">\n" +
+                            "        <EventTemplate><![CDATA[\n{\n" +
+                            "   \"@timestamp\": {\n" +
+                            "       \"$resolver\": \"timestamp\",\n" +
+                            "       \"pattern\": {\n" +
+                            "         \"format\": \"yyyy-MM-dd HH:mm:ss.SSSXXX\",\n" +
+                            "         \"timeZone\": \"UTC\"\n" +
+                            "        }\n" +
+                            "      },\n" +
+                            "      \"level\": {\n" +
+                            "       \"$resolver\": \"level\",\n" +
+                            "       \"field\": \"name\"\n" +
+                            "      },\n" +
+                            "      \"thread.name\": {\n" +
+                            "       \"$resolver\": \"thread\",\n" +
+                            "       \"field\": \"name\"\n" +
+                            "      },\n" +
+                            "      \"thread.id\": {\n" +
+                            "       \"$resolver\": \"thread\",\n" +
+                            "       \"field\": \"id\"\n" +
+                            "      },\n" +
+                            "      \"method\": {\n" +
+                            "       \"$resolver\": \"logger\",\n" +
+                            "       \"field\": \"name\",\n" +
+                            "       \"shorten\": {\n" +
+                            "         \"length\": 1,\n" +
+                            "         \"strategy\": \"tail\"\n" +
+                            "        }\n" +
+                            "      },\n" +
+                            "      \"line\": {\n" +
+                            "        \"$resolver\": \"source\",\n" +
+                            "        \"field\": \"lineNumber\"\n" +
+                            "      },\n" +
+                            "      \"message\": {\n" +
+                            "        \"$resolver\": \"message\",\n" +
+                            "        \"stringfield\": \"true\"\n" +
+                            "      },\n" +
+                            "      \"exception\": {\n" +
+                            "        \"$resolver\": \"exception\",\n" +
+                            "        \"field\": \"stackTrace\",\n" +
+                            "        \"stackTrace\": {\n" +
+                            "          \"stringified\": true,\n" +
+                            "          \"full\": true\n" +
+                            "        }\n" +
+                            "      }" +
+                            "      }\n]]></EventTemplate>\n" +
+                            "</JsonTemplateLayout>";
             properties.put("syslog_default_layout", jsonLayout);
             properties.put("syslog_warning_layout", jsonLayout);
             properties.put("syslog_audit_layout", jsonLayout);

--- a/fe/fe-core/src/main/java/com/starrocks/common/Log4jConfig.java
+++ b/fe/fe-core/src/main/java/com/starrocks/common/Log4jConfig.java
@@ -34,6 +34,7 @@
 
 package com.starrocks.common;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Maps;
 import groovy.lang.Tuple3;
@@ -61,12 +62,10 @@ public class Log4jConfig extends XmlConfiguration {
             "<Configuration status=\"info\" packages=\"com.starrocks.common\">\n" +
             "  <Appenders>\n" +
             "    <Console name=\"ConsoleErr\" target=\"SYSTEM_ERR\" follow=\"true\">\n" +
-            "      <PatternLayout pattern=\"%d{yyyy-MM-dd HH:mm:ss.SSSXXX} %p (%t|%tid) [%C{1}.%M():%L] %m%n\"/>\n" +
+            "      ${syslog_default_layout}\n" +
             "    </Console>\n" +
             "    <RollingFile name=\"Sys\" fileName=\"${sys_log_dir}/fe.log\" filePattern=\"${sys_log_dir}/fe.log.${sys_file_pattern}-%i\">\n" +
-            "      <PatternLayout charset=\"UTF-8\">\n" +
-            "        <Pattern>%d{yyyy-MM-dd HH:mm:ss.SSSXXX} %p (%t|%tid) [%C{1}.%M():%L] %m%n</Pattern>\n" +
-            "      </PatternLayout>\n" +
+            "      ${syslog_default_layout}\n" +
             "      <Policies>\n" +
             "        <TimeBasedTriggeringPolicy/>\n" +
             "        <SizeBasedTriggeringPolicy size=\"${sys_roll_maxsize}MB\"/>\n" +
@@ -79,9 +78,7 @@ public class Log4jConfig extends XmlConfiguration {
             "      </DefaultRolloverStrategy>\n" +
             "    </RollingFile>\n" +
             "    <RollingFile name=\"SysWF\" fileName=\"${sys_log_dir}/fe.warn.log\" filePattern=\"${sys_log_dir}/fe.warn.log.${sys_file_pattern}-%i\">\n" +
-            "      <PatternLayout charset=\"UTF-8\">\n" +
-            "        <Pattern>%d{yyyy-MM-dd HH:mm:ss.SSSXXX} %p (%t|%tid) [%C{1}.%M():%L] %m%n %ex</Pattern>\n" +
-            "      </PatternLayout>\n" +
+            "      ${syslog_warning_layout}\n" +
             "      <Policies>\n" +
             "        <TimeBasedTriggeringPolicy/>\n" +
             "        <SizeBasedTriggeringPolicy size=\"${sys_roll_maxsize}MB\"/>\n" +
@@ -94,9 +91,7 @@ public class Log4jConfig extends XmlConfiguration {
             "      </DefaultRolloverStrategy>\n" +
             "    </RollingFile>\n" +
             "    <RollingFile name=\"Auditfile\" fileName=\"${audit_log_dir}/fe.audit.log\" filePattern=\"${audit_log_dir}/fe.audit.log.${audit_file_pattern}-%i\">\n" +
-            "      <PatternLayout charset=\"UTF-8\">\n" +
-            "        <Pattern>%d{yyyy-MM-dd HH:mm:ss.SSSXXX} [%c{1}] %m%n</Pattern>\n" +
-            "      </PatternLayout>\n" +
+            "      ${syslog_audit_layout}\n" +
             "      <Policies>\n" +
             "        <TimeBasedTriggeringPolicy/>\n" +
             "        <SizeBasedTriggeringPolicy size=\"${audit_roll_maxsize}MB\"/>\n" +
@@ -109,9 +104,7 @@ public class Log4jConfig extends XmlConfiguration {
             "      </DefaultRolloverStrategy>\n" +
             "    </RollingFile>\n" +
             "    <RollingFile name=\"dumpFile\" fileName=\"${dump_log_dir}/fe.dump.log\" filePattern=\"${dump_log_dir}/fe.dump.log.${dump_file_pattern}-%i\">\n" +
-            "      <PatternLayout charset=\"UTF-8\">\n" +
-            "        <Pattern>%d{yyyy-MM-dd HH:mm:ss.SSSXXX} [%c{1}] %m%n</Pattern>\n" +
-            "      </PatternLayout>\n" +
+            "      ${syslog_dump_layout}\n" +
             "      <Policies>\n" +
             "        <TimeBasedTriggeringPolicy/>\n" +
             "        <SizeBasedTriggeringPolicy size=\"${dump_roll_maxsize}MB\"/>\n" +
@@ -124,9 +117,7 @@ public class Log4jConfig extends XmlConfiguration {
             "      </DefaultRolloverStrategy>\n" +
             "    </RollingFile>\n" +
             "    <RollingFile name=\"BigQueryFile\" fileName=\"${big_query_log_dir}/fe.big_query.log\" filePattern=\"${big_query_log_dir}/fe.big_query.log.${big_query_file_pattern}-%i\">\n" +
-            "      <PatternLayout charset=\"UTF-8\">\n" +
-            "        <Pattern>%d{yyyy-MM-dd HH:mm:ss.SSSXXX} [%c{1}] %m%n</Pattern>\n" +
-            "      </PatternLayout>\n" +
+            "      ${syslog_bigquery_layout}\n" +
             "      <Policies>\n" +
             "        <TimeBasedTriggeringPolicy/>\n" +
             "        <SizeBasedTriggeringPolicy size=\"${big_query_roll_maxsize}MB\"/>\n" +
@@ -139,9 +130,7 @@ public class Log4jConfig extends XmlConfiguration {
             "      </DefaultRolloverStrategy>\n" +
             "    </RollingFile>\n" +
             "    <RollingFile name=\"ProfileFile\" fileName=\"${profile_log_dir}/fe.profile.log\" filePattern=\"${profile_log_dir}/fe.profile.log.${profile_file_pattern}-%i\">\n" +
-            "      <PatternLayout charset=\"UTF-8\">\n" +
-            "        <Pattern>%d{yyyy-MM-dd HH:mm:ss.SSSXXX} [%c{1}] %m%n</Pattern>\n" +
-            "      </PatternLayout>\n" +
+            "      ${syslog_profile_layout}\n" +
             "      <Policies>\n" +
             "        <TimeBasedTriggeringPolicy/>\n" +
             "        <SizeBasedTriggeringPolicy size=\"${profile_log_roll_size_mb}MB\"/>\n" +
@@ -154,9 +143,7 @@ public class Log4jConfig extends XmlConfiguration {
             "      </DefaultRolloverStrategy>\n" +
             "    </RollingFile>\n" +
             "    <RollingFile name=\"InternalFile\" fileName=\"${internal_log_dir}/fe.internal.log\" filePattern=\"${internal_log_dir}/fe.internal.log.${internal_file_pattern}-%i\">\n" +
-            "      <PatternLayout charset=\"UTF-8\">\n" +
-            "        <Pattern>%d{yyyy-MM-dd HH:mm:ss.SSSXXX} %p (%t|%tid) [%C{1}.%M():%L] %m%n</Pattern>\n" +
-            "      </PatternLayout>\n" +
+            "      ${syslog_default_layout}\n" +
             "      <Policies>\n" +
             "        <TimeBasedTriggeringPolicy/>\n" +
             "        <SizeBasedTriggeringPolicy size=\"${internal_roll_maxsize}MB\"/>\n" +
@@ -211,7 +198,8 @@ public class Log4jConfig extends XmlConfiguration {
     private static String[] bigQueryModules;
     private static String[] internalModules;
 
-    private static void reconfig() throws IOException {
+    @VisibleForTesting
+    static String generateActiveLog4jXmlConfig() throws IOException {
         Map<String, String> properties = Maps.newHashMap();
 
         // sys log config
@@ -244,7 +232,7 @@ public class Log4jConfig extends XmlConfiguration {
         // big query log config
         properties.put("big_query_log_dir", Config.big_query_log_dir);
         properties.put("big_query_roll_maxsize", String.valueOf(Config.log_roll_size_mb));
-        properties.put("big_query_roll_num", String.valueOf(Config.big_query_log_roll_num));
+        properties.put("big_query_log_roll_num", String.valueOf(Config.big_query_log_roll_num));
         properties.put("big_query_log_delete_age", String.valueOf(Config.big_query_log_delete_age));
         properties.put("big_query_file_pattern",
                 getIntervalPattern("big_query_log_roll_interval", Config.big_query_log_roll_interval));
@@ -260,15 +248,46 @@ public class Log4jConfig extends XmlConfiguration {
         // internal log config
         properties.put("internal_log_dir", Config.internal_log_dir);
         properties.put("internal_roll_maxsize", String.valueOf(Config.log_roll_size_mb));
-        properties.put("internal_roll_num", String.valueOf(Config.internal_log_roll_num));
+        properties.put("internal_log_roll_num", String.valueOf(Config.internal_log_roll_num));
         properties.put("internal_log_delete_age", String.valueOf(Config.internal_log_delete_age));
         properties.put("internal_file_pattern",
                 getIntervalPattern("big_query_log_roll_interval", Config.internal_log_roll_interval));
 
+        // appender layout
+        final String jsonLoggingConfValue = "json";
+        if (jsonLoggingConfValue.equalsIgnoreCase(Config.sys_log_format)) {
+            // json logging
+            String jsonLayout =
+                    "<JSONLayout compact=\"true\" eventEol=\"true\" properties=\"true\" stacktraceAsString=\"true\" />";
+            properties.put("syslog_default_layout", jsonLayout);
+            properties.put("syslog_warning_layout", jsonLayout);
+            properties.put("syslog_audit_layout", jsonLayout);
+            properties.put("syslog_dump_layout", jsonLayout);
+            properties.put("syslog_bigquery_layout", jsonLayout);
+            properties.put("syslog_profile_layout", jsonLayout);
+        } else {
+            // fallback to plaintext logging
+            properties.put("syslog_default_layout",
+                    "<PatternLayout charset=\"UTF-8\" pattern=\"%d{yyyy-MM-dd HH:mm:ss.SSSXXX} %p (%t|%tid) [%C{1}.%M():%L] %m%n\"/>");
+            properties.put("syslog_warning_layout",
+                    "<PatternLayout charset=\"UTF-8\" pattern=\"%d{yyyy-MM-dd HH:mm:ss.SSSXXX} %p (%t|%tid) [%C{1}.%M():%L] %m%n %ex\"/>");
+            properties.put("syslog_audit_layout",
+                    "<PatternLayout charset=\"UTF-8\" pattern=\"%d{yyyy-MM-dd HH:mm:ss.SSSXXX} [%c{1}] %m%n\"/>");
+            properties.put("syslog_dump_layout",
+                    "<PatternLayout charset=\"UTF-8\" pattern=\"%d{yyyy-MM-dd HH:mm:ss.SSSXXX} [%c{1}] %m%n\"/>");
+            properties.put("syslog_bigquery_layout",
+                    "<PatternLayout charset=\"UTF-8\" pattern=\"%d{yyyy-MM-dd HH:mm:ss.SSSXXX} [%c{1}] %m%n\"/>");
+            properties.put("syslog_profile_layout",
+                    "<PatternLayout charset=\"UTF-8\" pattern=\"%d{yyyy-MM-dd HH:mm:ss.SSSXXX} [%c{1}] %m%n\"/>");
+        }
+
         String xmlConfTemplate = generateXmlConfTemplate();
         strSub = new StrSubstitutor(new Interpolator(properties));
-        xmlConfTemplate = strSub.replace(xmlConfTemplate);
+        return strSub.replace(xmlConfTemplate);
+    }
 
+    private static void reconfig() throws IOException {
+        String xmlConfTemplate = generateActiveLog4jXmlConfig();
         if (!FeConstants.runningUnitTest && !FeConstants.isReplayFromQueryDump) {
             System.out.println("=====");
             System.out.println(xmlConfTemplate);

--- a/fe/fe-core/src/test/java/com/starrocks/common/Log4jConfigTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/common/Log4jConfigTest.java
@@ -46,7 +46,7 @@ public class Log4jConfigTest {
         Config.sys_log_format = "plaintext";
         {
             String xmlConfig = Log4jConfig.generateActiveLog4jXmlConfig();
-            Assert.assertFalse(xmlConfig.contains("<JSONLayout"));
+            Assert.assertFalse(xmlConfig.contains("<JsonTemplateLayout"));
             Assert.assertTrue(xmlConfig.contains("<PatternLayout"));
             // no unresolved variable
             Matcher matcher = Pattern.compile(regStr, Pattern.MULTILINE).matcher(xmlConfig);
@@ -59,7 +59,7 @@ public class Log4jConfigTest {
         Config.sys_log_format = "json";
         {
             String xmlConfig = Log4jConfig.generateActiveLog4jXmlConfig();
-            Assert.assertTrue(xmlConfig.contains("<JSONLayout"));
+            Assert.assertTrue(xmlConfig.contains("<JsonTemplateLayout"));
             Assert.assertFalse(xmlConfig.contains("<PatternLayout"));
             // no unresolved variable
             Matcher matcher = Pattern.compile(regStr, Pattern.MULTILINE).matcher(xmlConfig);

--- a/fe/fe-core/src/test/java/com/starrocks/common/Log4jConfigTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/common/Log4jConfigTest.java
@@ -1,0 +1,72 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.common;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Log4jConfigTest {
+    private String logFormat;
+
+    @Before
+    public void setUp() throws IOException {
+        logFormat = Config.sys_log_format;
+        Log4jConfig.initLogging();
+    }
+
+    @After
+    public void tearDown() {
+        Config.sys_log_format = logFormat;
+    }
+
+    @Test
+    public void testJsonLoggingFormatConfig() throws IOException {
+        // ${([^}]*)}
+        String regStr = "\\$\\{([^\\}]*)\\}";
+        String name = "";
+        // default plaintext configuration
+        Config.sys_log_format = "plaintext";
+        {
+            String xmlConfig = Log4jConfig.generateActiveLog4jXmlConfig();
+            Assert.assertFalse(xmlConfig.contains("<JSONLayout"));
+            Assert.assertTrue(xmlConfig.contains("<PatternLayout"));
+            // no unresolved variable
+            Matcher matcher = Pattern.compile(regStr, Pattern.MULTILINE).matcher(xmlConfig);
+            if (matcher.find()) {
+                name = matcher.group(1);
+                Assert.fail(String.format("Unexpected of unresolved variables:'%s' in the final xmlConfig", name));
+            }
+        }
+        // check the json configuration
+        Config.sys_log_format = "json";
+        {
+            String xmlConfig = Log4jConfig.generateActiveLog4jXmlConfig();
+            Assert.assertTrue(xmlConfig.contains("<JSONLayout"));
+            Assert.assertFalse(xmlConfig.contains("<PatternLayout"));
+            // no unresolved variable
+            Matcher matcher = Pattern.compile(regStr, Pattern.MULTILINE).matcher(xmlConfig);
+            if (matcher.find()) {
+                name = matcher.group(1);
+                Assert.fail(String.format("Unexpected of unresolved variables:'%s' in the final xmlConfig", name));
+            }
+        }
+    }
+}

--- a/fe/pom.xml
+++ b/fe/pom.xml
@@ -405,6 +405,12 @@ under the License.
                 <version>${log4j.version}</version>
             </dependency>
 
+            <dependency>
+                <groupId>org.apache.logging.log4j</groupId>
+                <artifactId>log4j-layout-template-json</artifactId>
+                <version>${log4j.version}</version>
+            </dependency>
+
             <!-- https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j-1.2-api -->
             <dependency>
                 <groupId>org.apache.logging.log4j</groupId>

